Swap Subtree::token_trees from Vec to boxed slice

GnomedDev opened the pull request:
Performs one of the optimizations suggested in #16325, but a little bit more. Boxed slices guarantee `shrink_to_fit` aswell as saving a pointer width as no capacity has to be stored.  

Most of the diff is:
- Changing `vec![]` to `Box::new([])`
- Changing initialize -> fill into fill -> into_boxed_slice
- Working around the lack of an owned iterator or automatic iteration over a `Box<[T]>`

I would like to use my own crate, [small-fixed-array](https://lib.rs/small-fixed-array), although I understand if it isn't mature enough for this. If I'm given the go ahead, I can rework this PR to use it instead.

Veykril commented on crates/hir-expand/src/builtin_fn_macro.rs:
```diff
@@ -388,7 +393,12 @@ fn unreachable_expand(
+
+    // FIXME(slow): quote! have a way to expand to builder to make this a vec!
+    let mut mutable_trees = std::mem::take(&mut call.token_trees).into_vec();
+    mutable_trees.push(tt::TokenTree::Subtree(subtree));
```
same here

Veykril reviewed (commented):
Overall looks good (except for the `quote` re-allocs, but that isn't too troublesome I reckon). Would be good to do a quick perf measure by running analysis-stats on the repo (I can do that next week otherwise).

GnomedDev commented:
I've moved the push into a method on Subtree and added the reserve_exact. I'm not sure how to get stable results on `analysis-stats`, so it would probably be better for you to run it.

bors commented:
:pushpin: Commit a607e1b04b9263fa29e6000532e83f0526478440 has been approved by `lnicola`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=lnicola a607e1b04b9263fa29e6000532e83f0526478440 -->
<!-- homu: {"type":"Approved","sha":"a607e1b04b9263fa29e6000532e83f0526478440","approver":"lnicola","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:hourglass: Testing commit a607e1b04b9263fa29e6000532e83f0526478440 with merge 7fb639ffc18928cfc2fcdc4a0839b6e05bae4298...
<!-- homu: {"type":"BuildStarted","head_sha":"a607e1b04b9263fa29e6000532e83f0526478440","merge_sha":"7fb639ffc18928cfc2fcdc4a0839b6e05bae4298"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/7782103570/job/21218275976)
Approved by: lnicola
Pushing 7fb639ffc18928cfc2fcdc4a0839b6e05bae4298 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lnicola","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/7782103570/job/21218275976"},"merge_sha":"7fb639ffc18928cfc2fcdc4a0839b6e05bae4298"} -->
